pennylane_lightning/src/Gates.hpp

```diff
@@ -44,6 +44,16 @@ namespace Pennylane {
          * Generic matrix-multiplication kernel
          */
         virtual void applyKernel(const StateVector& state, const std::vector<size_t>& indices, const std::vector<size_t>& externalIndices, bool inverse);
+        
+        /**
+         * Generic matrix-multiplication kernel for applying the generator of an operation
+         */
+        void applyGenerator(const StateVector& state, const std::vector<size_t>& indices, const std::vector<size_t>& externalIndices);
```

Comment:
So the reason it's not `virtual void` is because we define `AbstractGate.applyGenerator()`, but just throw a `NotImplementedError` in case it's called?

Reply:
Actually a good question, we might need to make this `virtual`, see the [description here](https://stackoverflow.com/a/2392656). We will use `unique_ptr<AbstractGate>->applyGenerator` which can very well always call the base class method if it's not defined as `virtual`.